### ISSUE TITLE
Add the research tree GUI

### ISSUE THREAD

**Issue by Bertram25** [OpenDungeons/OpenDungeons#517]
Following #411 , the Research tree GUI and adjustments to the tab button and GUI is needed.
Here is the research tree first layout:
http://forum.freegamedev.net/viewtopic.php?p=62549#p62549
(The layout file is also available there.)
![research-tree-basic](https://cloud.githubusercontent.com/assets/765917/6252062/dcae03a0-b797-11e4-9502-66fa1ce0c650.png)

**How to adapt the OD/ImageButton:**
I'll be honest, I don't actually know yet. I guess a check mark or something should appear on the bottom right of the researches done, and something like a cog should appear on pending ones. The unavailable ones should simply appear disabled, even if you should be able to click on them to add them to the pending queue.

For that, I'll add new ImagerySection in ImageButton or maybe I'll duplicate that one and make something like a ResearchButton look'n'feel class:
https://github.com/OpenDungeons/OpenDungeons/blob/development/gui/OpenDungeons.looknfeel#L4091

Then, I'll need to have a look at how I can change the layers used on the different button states here depending on the research state: https://github.com/OpenDungeons/OpenDungeons/blob/development/gui/OpenDungeons.looknfeel#L4130

In the meanwhile, only available researches will be enabled, the rest (What's done and unreachable) will be disabled until I've figured out how to dynamically change the layers in a button state.

Useful link: http://static.cegui.org.uk/docs/0.8/classCEGUI_1_1StateImagery.html

**Comment by Bertram25** [comment]
I had an idea. I can add an empty layer for the standard button behaviour and ask to fill/empty them with an image from the code. :D It will be much simpler and can be used to show the pending/done states.

I still do think I'll overload the OD/ImageButton look'n'feel widget into a ResearchButton, but I'll add a label (text) support for both. Might be useful...

**Comment by hwoarangmy** [comment]
It could also be nice to display the research order somehow and to allow to edit it.
For that, you could display a number (1, 2, ...) on the researches telling which one is the next to be done. 1 would be the current one, 2 the next, ... But editing the order would not be straight forward.
Or you could add a "Pending researches" panel in the research tree and add the researches icons there when they are selected. Editing the order would be easier but it would need to add a new panel...

**Comment by Bertram25** [comment]
> It could also be nice to display the research order somehow and to allow to edit it.
> For that, you could display a number (1, 2, ...) on the researches telling which one is the next to be done. 

True, the image/research buttons will definitely need label support.

> But editing the order would not be straight forward.

To me (even if I do think we should do some trial and errors here), you'll first click on a research and a 1 appears bottom right of the button. You click on another a 2 appears, ...
If you want to change your queue, you click on a pending research (or even the one being researched atm) and the number of icon meaning it's being worked on disappears. All the other icons with a number **after** that one then gets decreased by one.
Thus, it's pretty straight-forward, and you _should_ not need a pending queue window/tab.
Let me add something in a PR and we'll discuss the bleeding edge of it while testing.

**Comment by hwoarangmy** [comment]
Yes, that's the behaviour I would expect as a player. But if you have configured the whole queue and, after scouting your opponent, you choose to change the research number 2 by a non selected one, you will need to redo all.
However, I'm not saying we shouldn't do that. I'm just pointing the weak point IMO of this logic. If we decide we can live with that, I'm ok. For my part, I can sleep even thinking that I will have to reconfigure the whole research list if I want to insert a new item in the pending list ^^

**Comment by Bertram25** [comment]
> However, I'm not saying we shouldn't do that. I'm just pointing the weak point IMO of this logic. If we decide we can live with that, I'm ok. For my part, I can sleep even thinking that I will have to reconfigure the whole research list if I want to insert a new item in the pending list ^^

In that sole case, I'd indeed add a clear research queue button. In fact, I'd add two. One to clear the queue, one to auto-select it (for new players).

**Comment by hwoarangmy** [comment]
ok. I guess it's the most simple and, after all, we can wait for players feedback to decide if we want to change that or not. OD is really becoming a game now ^^

**Comment by Bertram25** [comment]
Ok, part I of the evil plan has been merged with #521 ;)
I've made more visible icons for the main game resources, I hope you'll like them. (I do!)

Now that I've added new icons and unmixed the skin from the icons, I'll add the necessary GUI to  access settings and/or the research tree, objectives and future load/save features. That's the evil plan - Part II. :>

**Comment by Bertram25** [comment]
What is that?
![options](https://cloud.githubusercontent.com/assets/765917/6394749/f651b0d0-bdd3-11e4-81a4-ca11f00948ec.png)

More surprises tonight if everything goes well. ;)

**Comment by hwoarangmy** [comment]
Seems pretty nice ^^
Why are load game and save game disabled ?
BTW, Abort game should be quit game IMHO.

**Comment by Bertram25** [comment]
> Why are load game and save game disabled ?

Because it's just the GUI of it, the feature isn't done unfortunately. You might not notice it but I have worked on a button class that can display an icon next to the text, and I'm very happy about it. :D 
I also figured out how to add an state image and some text to what will be the research buttons. :)

> BTW, Abort game should be quit game IMHO.

Ok, I'll change that pre-PR.

**Comment by hwoarangmy** [comment]
> You might not notice it

Don't worry, it's pretty visible ^^ And very nice :p
For having fighted with CEGUI, I can guess how difficult it might have been to achieve that. Congrats.

> the feature isn't done unfortunately

You are right, I've not done the PR yet. Wait, have I said the feature works ????

**Comment by Bertram25** [comment]
> You are right, I've not done the PR yet. Wait, have I said the feature works ????

Stop teasing the teaser!! :P Anyway, the GUI overhaul should please everyone. :) We'll discuss the detail in the PR.

**Comment by hwoarangmy** [comment]
yep. I will wait for your PR to release mine because I've moved many stuff from Player to the Seat. So, our PRs will most likely conflict ^^
BTW, difference between Seat and Player was not very clear in my head. Now it is: Seat is for data that should be loaded/saved in level maps (owned gold, mana, researches, ...) and Player is for in game data. I had more or less naturally made the difference but now, it's clear.

**Comment by Bertram25** [comment]
> BTW, difference between Seat and Player was not very clear in my head. Now it is: Seat is for data that should be loaded/saved in level maps (owned gold, mana, researches, ...) and Player is for in game data. I had more or less naturally made the difference but now, it's clear.

:) If you happen to add that in the doc, i won't complain. 

> yep. I will wait for your PR to release mine because I've moved many stuff from Player to the Seat. So, our PRs will most likely conflict ^^

Sincerely, I do think not (yet). The next PR will be all about adding the GUI skeleton as neatly as possible. Then, I'll be on the logic.

**Comment by hwoarangmy** [comment]
ok. Then I will release the PR when I can. Saving will not be perfect (some states will not be saved like creature activity) but it will be more than enough ^^

**Comment by akien-mga** [comment]
For the 0.5.0 milestone, do you want to implement both the GUI and the final logic (i.e. the player can select the research they want to do, etc.), or do you prefer to start by giving a visual feedback of the current logic (i.e. you have the research GUI, but you can't "use" it to select research topics, you just get visual feedback on what is available, and what you are currently researching with the default implementation) ?

**Comment by Bertram25** [comment]
> For the 0.5.0 milestone, do you want to implement both the GUI and the final logic (i.e. the player can select the research they want to do, etc.), or do you prefer to start by giving a visual feedback of the current logic (i.e. you have the research GUI, but you can't "use" it to select research topics, you just get visual feedback on what is available, and what you are currently researching with the default implementation) ?

In two steps as you said. First step, plug the GUI and make the buttons update accordingly to the researches done, so that everyone can get to know how those research buttons work and have a function gui. The second step will be the full logic, but as I don't know at which point we'll release, let's say that either the first or both will be done for 0.5.0.

**Comment by Bertram25** [comment]
Note to self: Once this is merged and 0.5.0 is released, the new forge goes in the attack skills under the call to war spell.

**Comment by Bertram25** [comment]
Note to self 2: ~~The Imagebutton state icon should go over the disabled dim image.~~ At least I checked, it can't be done like that. Let's forget about that but I'll later have a look at colouring icons through parameters.

**Comment by Bertram25** [comment]
Note to self 3 : ~~The research state logic should be changed to use only one vector with each research states defined in it. I'll handle it.~~
I've tried it, it is working but isn't at all simpler since it involves a lot of fiddling when writing levels, ... Point dropped.

**Comment by Bertram25** [comment]
~~Arf, note to self4: Make the grimoire animated using the "Loop" animation.~~

**Comment by Bertram25** [comment]
Note to self 5 (It never stops!): ~~Make the tables rotated to 30° and adapt accordingly. Also, use a X/Y_SPOT_OFFSET constant as @hwoarangmy requested.~~ Done with #573

If everything goes fine, all that should be sorted out tonight.
I'll then move the rest of the research logic issue to 0.6.0 and focus on packaging stuff/marked tiles problems if @oyvindln doesn't do it before me.

**Comment by hwoarangmy** [comment]
@Bertram25 I will handle the grimoire animation as it is not really related to research GUI and it is easy ^^

**Comment by Bertram25** [comment]
> @Bertram25 I will handle the grimoire animation as it is not really related to research GUI and it is easy ^^

Fine to me :)

**Comment by Bertram25** [comment]
This issue is done. I'll open a new issue about the skill button. 
Unimplemented part supported by #840

**Comment by akien-mga** [comment]
Moving back to 0.5.0 milestone, actually the core of this issue was already fixed when 0.5.0 was released. Further improvements to the research logic were done in separate issues (like #817).
